### PR TITLE
feat(sdk): set job source in settings

### PR DESCRIPTION
**Pull request by TimH98**
Fixes
-----

Fixes [WB-13204](https://wandb.atlassian.net/browse/WB-13204)

Description
-----------
Allow users to set what type of job (`repo`, `artifact`, or `image`) to create when starting a run by setting `job_source` in Settings. Falls back to our usual logic if not specified.

Testing
-------
Some functional tests were modified to use the new `job_source` key.

### <samp>🤖 Generated by Copilot at a8ca029</samp>

> _To build a job from some source_
> _We added a new setting, of course_
> _The `job_source` attribute_
> _Can be set or implicit_
> _And it helps the `JobBuilder` enforce_

[WB-13204]: https://wandb.atlassian.net/browse/WB-13204?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by bcsherma — DISMISSED** [review]
looks good!

File: wandb/sdk/wandb_settings.py [hunk 1]
```diff
@@ -530,6 +530,7 @@ class Settings:
     ignore_globs: Tuple[str]
     init_timeout: float
     is_local: bool
+    job_source: str
```
Comment (dmitryduev) [inline wandb/sdk/wandb_settings.py:533]: Consider adding a validator to this setting that would check that it is one of the allowed types (look for _validator_... methods for examples on how we do it).

File: wandb/sdk/internal/job_builder.py [hunk 2]
```diff
@@ -80,6 +80,9 @@ def __init__(self, settings: SettingsStatic):
         self._summary = None
         self._logged_code_artifact = None
         self._disable = settings.disable_job_creation
+        self._source_type = None
```
Comment (dmitryduev) [inline wandb/sdk/internal/job_builder.py:83]: in here, do you want to make it an optional literal of the allowed strs?

**Review by dmitryduev — APPROVED** [review]
LGTM with a couple comments (main one is about setting validation)

**Review by moredatarequired — DISMISSED** [review]
Seems good to me!

**Review by KyleGoyette — APPROVED** [review]

**Pull request merged**